### DISCUSSION TITLE
Command mode configuration, accept shorthand suggestions

### DISCUSSION THREAD

**Discussion by russell-knight** [saghen/blink.cmp#744] [Q&A] [answered]
The main two commands I use are `:w` and `:q`. With blink autocomplete enabled for the command line when I type either of these I get a list of suggestions that includes the longform version of these commands e.g `write` present in the image below

![image](https://github.com/user-attachments/assets/f285efe6-ff37-4b97-a68e-26f57fe79399)

I can see that `wq` is also present in the list so perhaps there is a limit on command length imposed somewhere? I've tried adding a `min_keyword_length = 1` to `opts.sources`.

Is it possible to expose these shorthand commands I use in the autocomplete list? Or alternatively, is it possible to show the autocomplete list without _highlighting_ the first option, so that by typing `:w<CR>` I can save my file as usual?

**Comment by saghen** [comment, marked as answer]
You can make it not select by default in cmdline on v0.8.2 like https://cmp.saghen.dev/configuration/completion.html#list

The list itself comes from the built in completions so I'm not sure about adding shorthands, although I imagine it wouldn't be particularly useful to have single character options

**Reply by russell-knight** [reply]
This works perfectly thank you!

**Comment by gennaro-tedesco** [comment]
I have noticed that using the suggested
```lua
completion.list.selection = function(ctx)
  return ctx.mode == 'cmdline' and 'auto_insert' or 'preselect'
end
```
requires to press the accept twice for the command to be accepted in the commandline. In the recording below you can see that after selecting an entry (`help`) the first keypress only "closes" the completion window, whereas the second runs the command: is this intended behaviour?

![Kapture 2024-12-24 at 14 53 41](https://github.com/user-attachments/assets/612480ef-aafb-4d67-ae73-a6c7912d2a82)

My keymaps configuration is as follows:
```lua
keymap = {
	["<CR>"] = { "accept", "fallback" },
	["<Tab>"] = { "select_next", "fallback" },
	["<S-Tab>"] = { "select_prev", "fallback" },
	["<C-k>"] = { "scroll_documentation_up", "fallback" },
	["<C-j>"] = { "scroll_documentation_down", "fallback" },
	["<C-e>"] = { "hide" },
	["K"] = { "show", "show_documentation", "hide_documentation" },
}
```
and the same occurs where one uses `accept` or `select_and_accept`.

**Reply by norbertkeri** [reply]
I'm not the maintainer, but I was looking for a solution to exactly what you are describing, and there was a note somewhere in the docs that says you can provide a callback to accept:

```lua

            keymap = {
                cmdline = {
                    ['<cr>'] = {
                        function(cmp)
                            -- Enter should accept and execute (instead of just accept)
                            -- This lets me press enter once to execute a :cmdline command, instead of pressing twice
                            return cmp.accept({ callback = function() vim.api.nvim_feedkeys('\n', 'n', true) end })
                        end, "fallback"
                    },
                    ['<Tab>'] = { 'select_next' }
                },
         }
```

The tab binding is unrelated, I'm just leaving that in there because using keymap.cmdline removes all the default cmdline mappings, so you don't have a way to select the completions.

**Reply by gennaro-tedesco** [reply]
Thank you, I had missed such instruction, it works as intended indeed!
